[otbn] OpenTitan Big Number Accelerator Specification, v0.1

imphil opened the pull request:
Say hello to OTBN, the imaginative name for the OpenTitan Big Number
Accelerator. OTBN is a co-processor designed to quickly compute
algorithms which rely on wide integer arithmetic, primarily public key
cytography. Think RSA and ECDSA.

OTBN is a self-contained IP block and consists (on a very high level)
of a CPU, accompanied by a dedicated instruction and data memory.

The core of this specification is the instruction set architecture (ISA)
of OTBN. The ISA consists of two parts: a base instruction subset
for the control flow, and a big number instruction subset for wide
integer arithmetic.

The base instruction subset is based on an extended subset of RISC-V's
RV32I_Zisr. Even though many instructions are identical to their RV32I
couterpart, OTBN is *not* a standards-compliant RISC-V processor.

The big number instruction subset processes 256 bit wide data words.
It operates on a equally wide register file of 32 256b wide registers.
A lot of thought has been put into efficient multiplications of wide
integers, which resulted in the design of a 64 x 64 bit
multiply-accumulate instruction, which can be effectively combined to
multiply 256b words in 16 instructions.

A word of caution: The specification as it is presented in this commit
is not the "final" OTBN specification -- it is version 0.1 and marks a
starting point in the design of this IP block. The specification has
known holes, and has not yet incorporated all simplifications and
features we have thought of and require for the "final" OTBN. We
encourage ideas, suggestions, and critique in the form of issues, which
we can then sort through and incorporate into subsequent versions of
this specification.

Furthermore, the implementation work on OTBN has not yet started. We
expect to incorporate the learnings from our implementation, hardening,
and verification work to feed back into the specification.

The specification in this commit has been developed over the course of
more than half a year with a huge amount of work done by Felix Miller
with input from Greg Chadwick, Timothy Chen, Chris Gori, Miguel Osorio,
Stefan Wallentowitz, and many others. Thank you!

imphil commented:
This pull request is a markdown-ified version of the specification that we have created at https://docs.google.com/document/d/1PJmv_vYV_5ueFcM5D8sQjoHpbTYd6kTcNuZh87AOOSU/edit#. This specification has been reviewed and agreed on in the security meeting. If you have reviewed the Google Doc, you'll find a couple of clarifications, but no further changes.

This document is a first step towards a Big Number Accelerator OpenTitan might ship one day -- it is meant to be a living document for a while and serve as a starting point for an initial implementation draft.

At this point of the process I encourage reviewers to focus on the details and transcription errors of the Markdown version of the specification. Discussions going beyond that are very welcome, but best placed into either the enhancement proposals document at https://docs.google.com/document/u/3/d/1rgtPlXiJo21dq7gL_QZdV2XPISR78hkoLhJV8f1pt5M/edit, or in a new issue here on GitHub.

To view a HTML rendered version of this document, check out this branch on your local machine, and run `./util/build_docs.py --preview` where you cloned the opentitan repository. You can then point your browser to http://localhost:1313/hw/ip/otbn/doc/ to see the documentation.

imphil commented:
@msfschaffner @eunchan Especially with an eye on bronze, can you have a look at the interface in `otbn.hjson` in this PR and let me know if that's roughly the interface you'd expect for this IP block?

sjgitty commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2514 @@
+The design is kept as simple as possible to reduce the attack surface and aid verification and testing.
+For example, no interrupts or exceptions are included in the design, and all instructions are designed to be executable within a single cycle.
+
+OTBN is designed as a self-contained co-processor with its own instruction and data memory, which is accessible as bus device.
```
"accessible as *a* bus device"

sjgitty commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2514 @@
+    </td>
+  </tr>
+  <tr>
+    <td>...</td>
```
Since (it appears) only `x0` and `x1` are special, I recommend you add `x2` (General Purpose Register 2) prior to the `...` to show that the remaining 30 registers are GP

sjgitty commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2514 @@
+  </thead>
+  <tbody>
+    <tr>
+      <td>0x7C0</td>
```
are these `0x7c0` et al expected to be separated by 0x004 since they're all 32b wide? And I assume these `0x7c0` et al are not related to `x0 .. x31` correct? That table doesn't have a header row, so doesn't indicate `name` vs. this `number`, might be worth a header on that table.

imphil replied:
No relation to the GPRs. The table should have a header row (and has one in my rendering), let me know if your rendered version looks different.

![image](https://user-images.githubusercontent.com/1467123/84263462-71333400-ab17-11ea-9bd4-85aa31b0b5c0.png)

cdgori replied:
In the interest of clarity, should we put an explicit comment that there is no relationship between GPR and CSR?  (and perhaps between GPR and WDR too?)

sjgitty commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2514 @@
+    </tr>
+    <tr>
+      <td><code>BN.MULQACC.SO wr3.u, wr0.3, wr1.3, 0</code></td>
+      <td style="background-color: lightblue" colspan="2" rowspan="1">a3 * b3</code></td>
```
Missing one `<code>` here

felixmiller commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2514 @@
+    </tr>
+  </tbody>
+</table>
+
```
We might want to specify that # of iterations can't be zero. For the LOOPI instruction it might be debatable whether we want to support this. From a pure algorithmic point of view, to not loose generality, it would be better to have it. This would mean an instant jump to the end of the loop in this case. My expectation, however, would be that this creates a non-negligible HW overhead.

imphil replied:
Good point, as it uncovers a potential misunderstanding.

I thought of the iteration count as "how often is the code repeated", and not "how often is it executed". Meaning: the code in the loop body is always executed at least once.

If "loop count" is the "number of repeats", then setting it to 0 would mean execute the code coming next, but don't jump back to the beginning after the loop body ends.

Your suggestion to disallow 0 and my interpretation are probably identical, except for wasting one value in the immediate if we disallow "loop count == 0". I'd therefore change the spec to read something like:


```
LOOPI <repetitions>, <bodysize>
```

repetitions: Number of times the loop body is repeated.
repetition == 0 means the loop body is executed only once,  repetition == 1 means the loop body is executed twice, etc.

felixmiller commented:
> @felixmiller On CMPB/SUBB and flags. We currently specify for CMPB "The C flag retains its previous value if the value in wrs1 is equal to the value in wrs2.". For SUBB, I'd expect that the C flag is always set to either 0 or 1. Where is my thinking or the spec wrong?

The C and Z flags behave exactly the same for both the SUB(B) and CMP(B) instructions per the specification but I think we don't say what happens to the M and L flags. A CMP(B) instruction might set/reset those if the instruction uses the same logic as the SUB(B) instruction.
We can also leave this unspecified/implementation dependent but I think we should note that a SUB(B) instruction might modify the M and L flag and that it can't be assumed that e.g. a previously set L flag from an arithmetic operation survives a CMP(B) instruction.

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2525 @@
+
+### Loop Stack
+
+The LOOP instruction allows for nested loops; the active loops are stored on the loop stack.
```
I don't want to get into the implementation details in this spec, but should we indicate that CFI/other integrity checks can/will be applied to the loop stack?  (The only implementation detail that might be exposed here would be if we need to have some value-and-its-complement representation of the count, but we might formulate that internally as well.)

Also, since we make recommendations about GPR allocations for code authors in this document, should we make a recommendation that authors use the LOOP construct for data manipulation sequences?  (As opposed to the BEQ/BNE/JAL/JALR instructions?)  I am assuming that LOOP is going to be the most FI-robust structure we have, but perhaps that is not a correct assumption on my part.

imphil replied:
I know there's a lot more to be said (and discussed first) about how to best use OTBN in a secure fashion, but it certainly makes sense to include our current thinking into the spec already as "preliminary guidance".

@cdgori would you prefer such content to be part of the ISA spec, or would you look for it in a "(Secure) programmer's guide" section further down?

imphil replied:
I did file https://github.com/lowRISC/opentitan/issues/2499 and https://github.com/lowRISC/opentitan/issues/2498 to keep track of these issues. I'm not sure what we can add to the spec currently, as we want to make sure that we give the right advice, but we also know that there might be changes to the way loops works. If you're ok with it, I'd keep it as a TODO for now (and have added a comment to the doc mentioning that to our readers).

cdgori replied:
Thanks for filing the issues, mostly I was flagging things we need to remember to deal with as the spec/design matures.

Probably some "(secure) programmer's guide" will make sense ultimately, but it might be premature for us to write that yet.  Once we know what countermeasures/features we add it will be clear what the preferred coding styles are.

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2525 @@
+      <td>
+        <strong>MOD0</strong>.
+        Bits [31:0] of the modulus operand, used in the BN.ADDM/BN.SUBM instructions.
+        This CSR is mapped to the RND WSR.
```
This (and following 7 CSR) map to the MOD WSR, not the RND WSR, right?

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2525 @@
+      <td>R</td>
+      <td>
+        <strong>RND</strong>.
+        A random number.
```
Should we say explicitly this is different from (or related to) the RND WSR?

imphil replied:
Yes, we should say something. I'm just not quite sure what to say.

My thinking was the RND WSR and CSR are the same, i.e. a different view into the same LFSR/random number data output. This is effectively what we get in dcrypto, where there is only one RND data source used for control and data.

Obviously, only one register can be used at any given time, either through a 32b instruction, xor through a BN instruction.

Depending on the quality of the randomness, I can see reasons to use different random sources. We have not yet discussed the properties of RND any further, but there's an enhancement proposal tracking that at https://docs.google.com/document/d/1rgtPlXiJo21dq7gL_QZdV2XPISR78hkoLhJV8f1pt5M/edit#heading=h.k9tb5f5zh7q5.

For the purpose of pushing this document across the finish line, what's your thinking on that?

- Yes, the two RNDs can be the same and we don't need to discuss that much further. Let's put that statement into the v0.1 spec.
- Not sure, could be either way. Leave it unspecified or put a TODO in.
- They absolutely need to be different. Let's put that into the spec.

cdgori replied:
I think it's either the first or second option, so probably the second one for now, and if we find it's the first, we can clarify later.

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2525 @@
+      <td>RW</td>
+      <td>
+        <strong>MOD</strong>
+        Modulus. To be used in the BN.ADDM and BN.SUBM instructions.
```
Should probably note here as well the relationship to MOD0-7 CSR

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -0,0 +1,2525 @@
+      <td>R</td>
+      <td>
+        <strong>RND</strong>
+        A random number.
```
Should probably declare here the relationship or independence from the RND CSR.

imphil commented:
Thanks for your updates @felixmiller and @cdgori. I pushed an updated version.

Updates:
- Clarified the (non-) relation between GPRs, WDRs, and CSRs.
- Fixed description of MOD CSRs. They related to the MOD WSR, not the RND one.
- Mention the "reverse mapping" of the MOD WSR to the MOD0-7 CSRs.
- Added a warning to LOOP/LOOPI instructions that they are currently underspecified. We have already started looking into ways to specify them differently, e.g. by limiting the nesting and remove the full loop stack. I hope that we can still provide a consistent specification and improve the documentation of these instructions to be clearly understandable, but hold off on building too much advise on top of them, e.g. a detailed programmer guide.
- Modified the description of BN.CMP and BN.CMPB to mention BN.SUB/BN.SUBB and also align the implementation section. I have a feeling we'll go back to those instructions to 

Option questions:
- Questions around the relationship between the RND CSR and WSR. 
  We do have an enhancement proposal around this part of the spec in https://docs.google.com/document/d/1rgtPlXiJo21dq7gL_QZdV2XPISR78hkoLhJV8f1pt5M/edit#heading=h.k9tb5f5zh7q5. @cdgori: would you like to tackle the question of how the two RND special purpose registers relate (and what generally to expect from them) right now, or would you be OK with a TODO for now, which buys us a bit of time to discuss this question in detail while unblocking the first version of the spec?
- Questions around hardware loops. Since we don't have a good answer for many of them, and they depend on each other a bit, I have added a "warning note" into the spec to allow us to make forward progress. Option questions:
  - How to implement loops? I.e., do we want to keep the full loops stack?  https://github.com/lowRISC/opentitan/issues/2496
  - What kind of hardening can users expect?  https://github.com/lowRISC/opentitan/issues/2498
  - How to use them safely? https://github.com/lowRISC/opentitan/issues/2499
  - If we keep the current design: details around the range of immediates.

cdgori commented:
>@cdgori: would you like to tackle the question of how the two RND special purpose registers relate (and what generally to expect from them) right now, or would you be OK with a TODO for now, which buys us a bit of time to discuss this question in detail while unblocking the first version of the spec?

I think it's fine to mark this as TODO.

imphil commented on hw/ip/otbn/data/otbn.hjson:
```diff
@@ -0,0 +1,131 @@
+      hwqe: "true"
+      fields: [
+        { bits: "0"
+          name: "start"
```
Good point, I added a START_ADDR register.

imphil commented:
All review comments should now be addressed to the point that we can address them in a v0.1 spec. If nothing else comes up I'll merge this PR later tonight.

Thanks everyone for your feedback so far!
